azure/services/aso/service.go

```diff
@@ -83,6 +86,25 @@ func (s *Service[T, S]) Reconcile(ctx context.Context) error {
 		}
 	}
 
+	// If an error occurred during any create/update, don't delete any resources in case not all reconciled
+	// resources had the OwnerReference applied.
+	if (resultErr == nil || azure.IsOperationNotDoneError(resultErr)) && s.ListFunc != nil {
+		list, err := s.ListFunc(ctx, s.Scope.GetClient(), client.InNamespace(s.Scope.ASOOwner().GetNamespace()))
+		if err != nil {
+			resultErr = err
+		} else {
+			for _, existing := range list {
+				if _, exists := reconciled[existing.GetName()]; exists {
+					continue
+				}
+				err := s.Reconciler.DeleteResource(ctx, existing, s.Name())
```

Comment:
This `DeleteResource` will check for itself whether or not the ASO resource was managed by CAPZ and will skip the actual delete if it wasn't.